src/mavedb/lib/validation/keywords.py

```diff
@@ -16,30 +55,38 @@ def validate_keywords(keywords: list[str]):
     ValidationError
         If the list is invalid or null or if any individual keyword is invalid or null.
     """
-    if is_null(keywords):
+    if is_null(keyword) or not isinstance(keyword, str):
         raise ValidationError(
-            "{} are not valid keywords. Keywords must be a non null list of strings.".format(keywords)
+            "{} are not valid keywords. Keywords must be a non null list of strings.".format(keyword)
         )
-    else:
-        for keyword in keywords:
-            validate_keyword(keyword)
 
 
-def validate_keyword(keyword: str):
-    """
-    This function validates whether or not the kw parameter is valid by
-    checking that it is a string that is not null. If kw is null
-    or is not a string, an error is raised.
+def validate_keyword_keys(keywords: list):
+    keys = []
+    values = []
+    for k in keywords:
+        keys.append(k.keyword.key.lower())
+        values.append(k.keyword.value.lower())
 
-    Parameters
-    __________
-    kw : str
-        The keyword to be validated.
+    if "endogenous locus library method" in values:
+        if ("endogenous locus library method system" not in keys) or ("endogenous locus library method mechanism" not in keys):
+            raise ValidationError("Miss 'Endogenous Locus Library Method System' "
+                                  "or 'Endogenous Locus Library Method Mechanism' in keywords")
+        elif ("in vitro construct library method system" in keys) or ("in vitro construct library method mechanism" in keys):
+            raise ValidationError("Endogenous Locus Library Method does not allow in vitro method system or mechanism")
+    elif "in vitro construct library method" in values:
+        if ("in vitro construct library method system" not in keys) or ("in vitro construct library method mechanism" not in keys):
+            raise ValidationError("Miss 'In Vitro Construct Library Method System' "
+                                  "or 'In Vitro Construct Library Method Mechanism' in keywords")
+        elif ("endogenous locus library method system" in keys) or ("endogenous locus library method mechanism" in keys):
+            raise ValidationError(
+                "In Vitro Construct Library Method does not allow endogenous method system or mechanism")
+    elif "other" in values:
```

Comment:
It seems like this will cause problems if we allow `other` as a value for any other key in the future. It seems to me that it would be better to look at particular key-value combinations rather than looking at keys and values as separate lists. Since only one value is allowed per key, one way to do this is to turn the keywords list into a dictionary.

Reply:
Thanks Jeremy! This validator should only work for the first keyword 'Variant Library Creation Method' which has these three values and shall not affect other keys that users choose 'Other'. I'll modify it. I'm wondering if it's necessary to check 'endogenous locus library method system' and 'endogenous locus library method mechanism' when users select 'Endogenous locus library method.' The same applies to 'in vitro construct library method.' For example, users only choose either 'Endogenous locus library method' or 'In vitro construct library method' in 'Variant Library Creation Method', but don't want to choose anything in other keywords.